### PR TITLE
AO3-4934 Extend kudos tests, fix error posting kudos on non-existent works

### PR DESCRIPTION
**Pull request by hatal175**
## Issue

https://otwarchive.atlassian.net/browse/AO3-4934

## Purpose

Completed coverage for the kudos controller

## Testing

No new features just tests. Run them :)

## Credit

Tal Hayon

Please use he

### CODE REVIEW

**Comment by hatal175** [comment]
Please do not bother reviewing until I've fixed the tests.

File: app/controllers/kudos_controller.rb [hunk 1]
```diff
@@ -23,7 +23,7 @@ def create
         format.html do
           flash[:comment_notice] = ts("Thank you for leaving kudos!")
 
-          redirect_to request.referer and return
+          redirect_to request.referer || root_path and return
```
Comment (sarken) [inline app/controllers/kudos_controller.rb]: I'm not sure root_path is a good fallback for kudos -- if we can't redirect someone back to the _exact_ referrer (which could be something like a specific chapter of the work they're kudosing or the work's kudos), we should at least redirect them to work.

File: app/sweepers/kudos_sweeper.rb [hunk 2]
```diff
@@ -11,7 +11,7 @@ def after_create(kudo)
 
     # expire the cache for the kudos section in the view
     # and for the full-page, expanded list
-    expire_fragment "#{kudo.commentable.cache_key}/kudos-v1"
-    expire_fragment "#{kudo.commentable.cache_key}/kudos/full"
+    ActionController::Base.new.expire_fragment("#{kudo.commentable.cache_key}/kudos-v1")
```
[the remark this answers is not in the document]
Comment (redsummernight) [reply app/sweepers/kudos_sweeper.rb]: It won't fix that unfortunately. This is for a different issue.

If we use plain `expire_fragment` for this PR, the errors we get are:
```
  1) ChaptersController show assigns @kudos to non-anonymous kudos
     Failure/Error: expire_fragment "#{kudo.commentable.cache_key}/kudos-v1"
     
     NoMethodError:
       undefined method `expire_fragment' for #<KudosSweeper:0x000055cb65a32718 @controller=nil>
     # ./app/sweepers/kudos_sweeper.rb:14:in `after_create'
     # ./spec/controllers/chapters_controller_spec.rb:199:in `block (3 levels) in <top (required)>'

  2) ChaptersController show assigns instance variables correctly
     Failure/Error: expire_fragment "#{kudo.commentable.cache_key}/kudos-v1"
     
     NoMethodError:
       undefined method `expire_fragment' for #<KudosSweeper:0x000055cb65a32718 @controller=nil>
     # ./app/sweepers/kudos_sweeper.rb:14:in `after_create'
     # ./spec/controllers/chapters_controller_spec.rb:209:in `block (3 levels) in <top (required)>'
```

Totally out of the left field. I think what happened was:

- There's a bug when using sweepers with Rails 4 and RSpec controller tests: `https://github.com/rails/rails-observers/issues/31` (no cross-repo linking please, GitHub).
- A workaround is to use `ActionController::Base.new.expire_fragment` instead of just `expire_fragment`. We fixed this for the other sweepers in the repo (#2917), so why not this one?
- We never have any RSpec coverage for the kudos controller until now, which means finally this line gets run in specs, revealing the error: https://github.com/otwcode/otwarchive/blob/e59b8eba05cb6630b6a01f956be74076d64291fa/app/controllers/kudos_controller.rb#L3

In conclusion, :joy:.

[the remark this answers is not in the document]
Comment (sarken) [reply app/sweepers/kudos_sweeper.rb]: You say 😂, I say, "I wish :sob_blood: were a default emoji."

File: app/controllers/kudos_controller.rb [hunk 3]
```diff
@@ -44,11 +52,11 @@ def create
           if @kudo && @kudo.creator_of_work?
             error_message = "You can't leave kudos on your own work."
           end
-          if !current_user.present? && commentable.restricted?
+          if !current_user.present? && !commentable.nil? && commentable.restricted?
```
[the remark this answers is not in the document]
Comment (redsummernight) [reply app/controllers/kudos_controller.rb]: I was gonna make this PR test-only but I had to fix the sweepers anyway. Should I fix this too?

[the remark this answers is not in the document]
Comment (sarken) [reply app/controllers/kudos_controller.rb]: Yeah, if it'll work, might as well. Thanks!

File: spec/controllers/kudos_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  include RedirectExpectationHelper
+
+  describe "create" do
+    context "when work is public" do
+      let(:work) { create(:posted_work) }
+
+      context "when kudos are given from work" do
+        it "redirects to referer with a notice" do
+          referer = work_path(work)
+          request.headers["HTTP_REFERER"] = referer
```
Comment (redsummernight) [inline spec/controllers/kudos_controller_spec.rb]: This is the (recently) recommended way of setting request headers in controller specs:

    https://github.com/rspec/rspec-rails/pull/2229

It looks marginally cleaner than:

https://github.com/otwcode/otwarchive/blob/540f520aef15d32d57a85e5c9751a2924cea7fa5/spec/controllers/inbox_controller_spec.rb#L136-L142

File: app/controllers/kudos_controller.rb (hunk 1 above)
Comment (redsummernight) [reply app/controllers/kudos_controller.rb]: I reverted this change.
